ragna/deploy/_ui/components/metadata_filters_builder.py

```diff
@@ -0,0 +1,248 @@
+            ]
+
+    def did_click_add_filter_row_button(self, event):
+        if not self.filter_rows[-1].is_empty():
+            new_filter_row = FilterRow(
+                valid_key_value_pairs=self.corpus_metadata[
+                    self.corpus_names_select.value
+                ],
+                key="",
+                operator="",
+                value="",
+                on_delete_callback=self.delete_filter_row,
+            )
+            self.filter_rows = self.filter_rows + [new_filter_row]
+
+    def delete_filter_row(self, event):
+        filter_row_to_remove = None
+        for filter_row in self.filter_rows:
+            if event.obj == filter_row.delete_button:
+                filter_row_to_remove = filter_row
+                break
+
+        if filter_row_to_remove is None:
+            return
+
+        new_filter_rows = [f for f in self.filter_rows if f != filter_row_to_remove]
+        if len(new_filter_rows) == 0:
+            new_filter_rows.append(
+                FilterRow(
+                    valid_key_value_pairs=self.corpus_metadata[
+                        self.corpus_names_select.value
+                    ],
+                    key="",
+                    operator="",
+                    value="",
+                    on_delete_callback=self.delete_filter_row,
+                )
+            )
+
+        self.filter_rows = new_filter_rows
+
+    @pn.depends("filter_rows")
+    def render_filters_rows(self):
+        return pn.Column(
+            *self.filter_rows,
+            css_classes=["metadata-filter-collection"],
+        )
+
+    def construct_metadata_filters(self):
+        metadata_filters = [
+            filter_row.construct_metadata_filter()
+            for filter_row in self.filter_rows
+            if filter_row.construct_metadata_filter() is not None
+        ]
+
+        combined_metadata_filters = MetadataFilter.and_(metadata_filters).to_primitive()
+
+        if not combined_metadata_filters:
+            return None
+        else:
+            return combined_metadata_filters
+
+    def validate(self):
+        result = True
+        for filter_row in self.filter_rows:
+            # If the last filter is empty, we do not want to validate it.
+            # This allows to have only one empty filter, to question on the whole corpus.
+            if filter_row == self.filter_rows[-1] and filter_row.is_empty():
+                continue
+
+            if not filter_row.validate():
+                result = False
+                # Do not break, we want to call validate() on every filters
+
+        return result
+
+    def __panel__(self):
+        return pn.Column(
+            pn.pane.HTML("<b>Available Corpuses</b>"),
+            self.corpus_names_select,
+            pn.pane.HTML("<b>Metadata Filters</b>"),
+            pn.Column(
+                self.render_filters_rows,
+                pn.Row(self.add_filter_row_button),
+                max_height=160,
```

Comment:
Restrict the height to 4 rows of metadata filters, and then scroll will be introduced along y-axis (see CSS).